File: src/main/java/com/calendar/Event.java
package com.calendar;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.UUID;

public class Event {

    private String id;
    private String title;
    private String description;
    private LocalDate date;
    private LocalTime startTime;
    private LocalTime endTime;
    private String color;

    public Event() {
        this.id = UUID.randomUUID().toString().substring(0, 8);
        this.color = "#1a73e8";
    }

    public Event(String title, String description, LocalDate date,
                 LocalTime startTime, LocalTime endTime, String color) {
        this();
        this.title = title;
        this.description = description;
        this.date = date;
        this.startTime = startTime;
        this.endTime = endTime;
        if (color != null && !color.isBlank()) {
            this.color = color;
        }
    }

    public String getId() { return id; }
    public void setId(String id) { this.id = id; }

    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public LocalDate getDate() { return date; }
    public void setDate(LocalDate date) { this.date = date; }

    public LocalTime getStartTime() { return startTime; }
    public void setStartTime(LocalTime startTime) { this.startTime = startTime; }

    public LocalTime getEndTime() { return endTime; }
    public void setEndTime(LocalTime endTime) { this.endTime = endTime; }

    public String getColor() { return color; }
    public void setColor(String color) { this.color = color; }

    public String getTimeRange() {
        if (startTime == null) return "";
        String start = formatTime(startTime);
        if (endTime == null) return start;
        return start + " – " + formatTime(endTime);
    }

    private String formatTime(LocalTime time) {
        int hour = time.getHour();
        int minute = time.getMinute();
        String amPm = hour < 12 ? "AM" : "PM";
        int displayHour = hour % 12;
        if (displayHour == 0) displayHour = 12;
        if (minute == 0) return displayHour + " " + amPm;
        return displayHour + ":" + String.format("%02d", minute) + " " + amPm;
    }
}

File: src/main/kotlin/com/calendar/Event.kt
package com.calendar

import com.fasterxml.jackson.annotation.JsonProperty
import java.time.LocalDate
import java.time.LocalTime
import java.util.UUID

class Event {

    // Review fix: id is read-only from JSON — server always generates it
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    var id: String = UUID.randomUUID().toString()
        private set

    var title: String? = null
    var description: String? = null
    var date: LocalDate? = null
    var startTime: LocalTime? = null
    var endTime: LocalTime? = null
    var color: String = DEFAULT_COLOR

    constructor()

    constructor(
        title: String,
        description: String?,
        date: LocalDate,
        startTime: LocalTime?,
        endTime: LocalTime?,
        color: String?
    ) {
        this.title = title
        this.description = description
        this.date = date
        this.startTime = startTime
        this.endTime = endTime
        if (!color.isNullOrBlank()) {
            this.color = color
        }
    }

    /**
     * Regenerate the id with a fresh full UUID.
     * Called by EventService.create to ensure server-controlled ids.
     */
    fun regenerateId() {
        this.id = UUID.randomUUID().toString()
    }

    val timeRange: String
        get() {
            val start = startTime ?: return ""
            val startStr = formatTime(start)
            val end = endTime ?: return startStr
            return "$startStr \u2013 ${formatTime(end)}"
        }

    private fun formatTime(time: LocalTime): String {
        val h = time.hour
        val m = time.minute
        val amPm = if (h < 12) "AM" else "PM"
        val displayHour = if (h % 12 == 0) 12 else h % 12
        return if (m == 0) "$displayHour $amPm"
        else "$displayHour:${"%02d".format(m)} $amPm"
    }

    companion object {
        const val DEFAULT_COLOR = "#1a73e8"
        val COLOR_PATTERN = Regex("^#[0-9a-fA-F]{3}([0-9a-fA-F]{3})?\$")
    }
}
